packages/@lwc/engine-core/src/framework/wiring.ts

```diff
@@ -73,92 +80,219 @@ function createMethodDataCallback(vm: VM, method: (data: any) => any) {
     };
 }
 
-function createConfigWatcher(
-    component: ComponentInterface,
-    configCallback: ConfigCallback,
-    callbackWhenConfigIsReady: (newConfig: ConfigValue) => void
-): { computeConfigAndUpdate: () => void; ro: ReactiveObserver } {
-    let hasPendingConfig: boolean = false;
-    // creating the reactive observer for reactive params when needed
-    const ro = new ReactiveObserver(() => {
-        if (hasPendingConfig === false) {
-            hasPendingConfig = true;
-            // collect new config in the micro-task
+interface WireAdapterDecorator extends WireAdapter {
+    computeConfig: () => ConfigValue;
+}
+
+class BaseWireAdapter implements WireAdapterDecorator {
+    private readonly vm;
+    private readonly connector;
+    private readonly configCallback;
+    private readonly component;
+
+    constructor(connector: WireAdapter, vm: VM, configCallback: ConfigCallback) {
+        this.connector = connector;
+        this.vm = vm;
+        this.component = vm.component;
+        this.configCallback = configCallback;
+    }
+
+    computeConfig() {
+        return this.configCallback(this.component);
+    }
+
+    update(config: ConfigValue, context?: ContextValue) {
+        runWithBoundaryProtection(
+            this.vm,
+            this.vm,
+            noop,
+            () => {
+                // job
+                this.connector.update(config, context);
+            },
+            noop
+        );
+    }
+
+    connect() {
+        runWithBoundaryProtection(
+            this.vm,
+            this.vm,
+            noop,
+            () => {
+                // job
+                this.connector.connect();
+            },
+            noop
+        );
+    }
+
+    disconnect() {
+        runWithBoundaryProtection(
+            this.vm,
+            this.vm,
+            noop,
+            () => {
+                // job
+                this.connector.disconnect();
+            },
+            noop
+        );
+    }
+}
+
+class ContextAwareWireAdapter implements WireAdapterDecorator {
+    private readonly decoratedWireAdapter;
+    private readonly vm;
+    private readonly adapterContextToken;
+    private readonly renderer;
+    private readonly elm;
+
+    private context: ContextValue | undefined;
+    private wiredDisconnecting: Array<() => void> = [];
+
+    private setNewContext = (newContext: ContextValue) => {
+        // eslint-disable-next-line lwc-internal/no-invalid-todo
+        // TODO: dev-mode validation of config based on the adapter.contextSchema
+        if (this.context !== newContext) {
```

Comment:
`this` in an arrow function set into a private field seems like the wrong thing to do.